drivers/dma/dma_dw_common.c

```diff
@@ -475,15 +475,17 @@ int dw_dma_start(const struct device *dev, uint32_t channel)
 	dw_write(dev_cfg->base, DW_LLP(channel), llp);
 	LOG_DBG("ctrl_lo %x, masked ctrl_lo %x, LLP %x",
 		lli->ctrl_lo, masked_ctrl_lo, dw_read(dev_cfg->base, DW_LLP(channel)));
-#else
+#endif /* CONFIG_DMA_DW_HW_LLI */
+
+#if !CONFIG_DMA_DW_HW_LLI || CONFIG_SOC_SERIES_INTEL_ADSP_CAVS
 	/* channel needs to start from scratch, so write SAR and DAR */
 	dw_write(dev_cfg->base, DW_SAR(channel), lli->sar);
 	dw_write(dev_cfg->base, DW_DAR(channel), lli->dar);
 
 	/* program CTL_LO and CTL_HI */
 	dw_write(dev_cfg->base, DW_CTRL_LOW(channel), lli->ctrl_lo);
 	dw_write(dev_cfg->base, DW_CTRL_HIGH(channel), lli->ctrl_hi);
-#endif /* CONFIG_DMA_DW_HW_LLI */
+#endif
```

Comment:
Interesting, with the said patch the CTRL_LO (and _HI) write is skipped in case of LLP/LLI use, but to use the linked list mode either or both of DW_CTLL_LLP_D_EN, DW_CTLL_LLP_S_EN should be set.
Since we want both SAR and DAR to be taken from the LI descriptor in any case and not to write directly to the SAR/DAR we need both flags set unconditionally?

I think the correct fix would be:
```diff
diff --git a/drivers/dma/dma_dw_common.c b/drivers/dma/dma_dw_common.c
index 16fd3c875b8f..9d84f236ad5c 100644
--- a/drivers/dma/dma_dw_common.c
+++ b/drivers/dma/dma_dw_common.c
@@ -278,19 +278,11 @@ int dw_dma_config(const struct device *dev, uint32_t channel,
 		case MEMORY_TO_MEMORY:
 			lli_desc->ctrl_lo |= DW_CTLL_FC_M2M | DW_CTLL_SRC_INC |
 				DW_CTLL_DST_INC;
-#if CONFIG_DMA_DW_HW_LLI
-			LOG_DBG("setting LLP_D_EN, LLP_S_EN in lli_desc->ctrl_lo %x",
-				lli_desc->ctrl_lo);
-			lli_desc->ctrl_lo |=
-				DW_CTLL_LLP_S_EN | DW_CTLL_LLP_D_EN;
-			LOG_DBG("lli_desc->ctrl_lo %x", lli_desc->ctrl_lo);
-#endif
 			break;
 		case MEMORY_TO_PERIPHERAL:
 			lli_desc->ctrl_lo |= DW_CTLL_FC_M2P | DW_CTLL_SRC_INC |
 				DW_CTLL_DST_FIX;
 #if CONFIG_DMA_DW_HW_LLI
-			lli_desc->ctrl_lo |= DW_CTLL_LLP_S_EN;
 			chan_data->cfg_lo |= DW_CFGL_RELOAD_DST;
 #endif
 			/* Assign a hardware handshake interface (0-15) to the
@@ -302,9 +294,7 @@ int dw_dma_config(const struct device *dev, uint32_t channel,
 			lli_desc->ctrl_lo |= DW_CTLL_FC_P2M | DW_CTLL_SRC_FIX |
 				DW_CTLL_DST_INC;
 #if CONFIG_DMA_DW_HW_LLI
-			if (!block_cfg->dest_scatter_en) {
-				lli_desc->ctrl_lo |= DW_CTLL_LLP_D_EN;
-			} else {
+			if (block_cfg->dest_scatter_en) {
 				/* Use contiguous auto-reload. Line 3 in
 				 * table 3-3
 				 */
@@ -360,7 +350,7 @@ int dw_dma_config(const struct device *dev, uint32_t channel,
 	}
 
 #if CONFIG_DMA_DW_HW_LLI
-	chan_data->cfg_lo |= DW_CFGL_CTL_HI_UPD_EN;
+	chan_data->cfg_lo |= DW_CFGL_CTL_HI_UPD_EN | DW_CTLL_LLP_D_EN | DW_CTLL_LLP_S_EN;
 #endif
 
 	/* end of list or cyclic buffer */
@@ -479,11 +469,11 @@ int dw_dma_start(const struct device *dev, uint32_t channel)
 	/* channel needs to start from scratch, so write SAR and DAR */
 	dw_write(dev_cfg->base, DW_SAR(channel), lli->sar);
 	dw_write(dev_cfg->base, DW_DAR(channel), lli->dar);
+#endif /* CONFIG_DMA_DW_HW_LLI */
 
 	/* program CTL_LO and CTL_HI */
 	dw_write(dev_cfg->base, DW_CTRL_LOW(channel), lli->ctrl_lo);
 	dw_write(dev_cfg->base, DW_CTRL_HIGH(channel), lli->ctrl_hi);
-#endif /* CONFIG_DMA_DW_HW_LLI */
 
 	/* program CFG_LO and CFG_HI */
 	dw_write(dev_cfg->base, DW_CFG_LOW(channel), chan_data->cfg_lo);
```

Reply:
This doesn't seem quite right, LLP bits are part of the ctl_lo register not cfg_lo register.

Reply:
revisiting the `diff` variations with another one which was tested on MTL at least:
```diff
diff --git a/drivers/dma/dma_dw_common.c b/drivers/dma/dma_dw_common.c
index 16fd3c875b8f..8003330e22f1 100644
--- a/drivers/dma/dma_dw_common.c
+++ b/drivers/dma/dma_dw_common.c
@@ -475,15 +475,20 @@ int dw_dma_start(const struct device *dev, uint32_t channel)
 	dw_write(dev_cfg->base, DW_LLP(channel), llp);
 	LOG_DBG("ctrl_lo %x, masked ctrl_lo %x, LLP %x",
 		lli->ctrl_lo, masked_ctrl_lo, dw_read(dev_cfg->base, DW_LLP(channel)));
-#else
-	/* channel needs to start from scratch, so write SAR and DAR */
-	dw_write(dev_cfg->base, DW_SAR(channel), lli->sar);
-	dw_write(dev_cfg->base, DW_DAR(channel), lli->dar);
+#endif /* CONFIG_DMA_DW_HW_LLI */
+
+	/* Program the addresses only if they are not taken from LLI
+	 * DW_CFGL_RELOAD_SRC/DW_CFGL_RELOAD_DST only set in case of LLI and for the side which is
+	 * constant, not updated from the descriptor
+	 */
+	if (chan_data->cfg_lo & DW_CFGL_RELOAD_SRC)
+		dw_write(dev_cfg->base, DW_SAR(channel), lli->sar);
+	if (chan_data->cfg_lo & DW_CFGL_RELOAD_DST)
+		dw_write(dev_cfg->base, DW_DAR(channel), lli->dar);
 
 	/* program CTL_LO and CTL_HI */
 	dw_write(dev_cfg->base, DW_CTRL_LOW(channel), lli->ctrl_lo);
 	dw_write(dev_cfg->base, DW_CTRL_HIGH(channel), lli->ctrl_hi);
-#endif /* CONFIG_DMA_DW_HW_LLI */
 
 	/* program CFG_LO and CFG_HI */
 	dw_write(dev_cfg->base, DW_CFG_LOW(channel), chan_data->cfg_lo);
```
The CTL_LO/HI must be programmed unconditionally, but the SAR/DAR only need to be set in register if it is not taken from LLI.

In theory this is correct.